src/transformers/models/qwen2_vl/modeling_qwen2_vl.py

```diff
@@ -0,0 +1,1653 @@
+
+            If `past_key_values` are used, the user can optionally input only the last `decoder_input_ids` (those that
+            don't have their past key value states given to this model) of shape `(batch_size, 1)` instead of all
+            `decoder_input_ids` of shape `(batch_size, sequence_length)`.
+        inputs_embeds (`torch.FloatTensor` of shape `(batch_size, sequence_length, hidden_size)`, *optional*):
+            Optionally, instead of passing `input_ids` you can choose to directly pass an embedded representation. This
+            is useful if you want more control over how to convert `input_ids` indices into associated vectors than the
+            model's internal embedding lookup matrix.
+        use_cache (`bool`, *optional*):
+            If set to `True`, `past_key_values` key value states are returned and can be used to speed up decoding (see
+            `past_key_values`).
+        output_attentions (`bool`, *optional*):
+            Whether or not to return the attentions tensors of all attention layers. See `attentions` under returned
+            tensors for more detail.
+        output_hidden_states (`bool`, *optional*):
+            Whether or not to return the hidden states of all layers. See `hidden_states` under returned tensors for
+            more detail.
+        return_dict (`bool`, *optional*):
+            Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
+        pixel_values (`torch.FloatTensor` of shape `(seq_length, num_channels * image_size * image_size)):
+            The tensors corresponding to the input images. Pixel values can be obtained using
+            [`AutoImageProcessor`]. See [`Qwen2VLImageProcessor.__call__`] for details. [`Qwen2VLProcessor`] uses
+            [`Qwen2VLImageProcessor`] for processing images.
+        pixel_values_videos (`torch.FloatTensor` of shape `(seq_length, num_channels * temporal_size * image_size * image_size)):
+            The tensors corresponding to the input videos. Pixel values can be obtained using
+            [`AutoImageProcessor`]. See [`Qwen2VLImageProcessor.__call__`] for details. [`Qwen2VLProcessor`] uses
+            [`Qwen2VLImageProcessor`] for processing videos.
+        image_grid_thw (`torch.LongTensor` of shape `(num_images, 3)`, *optional*):
+            The temporal, height and width of feature shape of each image in LLM.
+        video_grid_thw (`torch.LongTensor` of shape `(num_videos, 3)`, *optional*):
+            The temporal, height and width of feature shape of each video in LLM.
+        rope_deltas (`torch.LongTensor` of shape `(batch_size, )`, *optional*):
+            The rope index difference between sequence length and multimodal rope.
+"""
+
+
+class Qwen2VLForConditionalGeneration(Qwen2VLPreTrainedModel):
+    _tied_weights_keys = ["lm_head.weight"]
+
+    def __init__(self, config):
+        super().__init__(config)
+        self.visual = Qwen2VisionTransformer(config.vision_config)
```

Comment:
In continuation of the attn implementation discussion from above. After we get text and vision configs, we can do  `Qwen2VisionTransformer._from_config(config.vision_config, attn_implementation=config._attn_implementation)`. The vision block will dispatch on the requested attention, if it is available. Same for LM bacbone, `Qwen2VLModel._from_config(config, attn_implementation=config._attn_implementation)`

The only thing is that `Qwen2VisionTransformer` should also be a QwenVLMPreTrainedModel to be able to dispatch automatically different attentions. And that means we'll need a proper init and forward docs for the VisionTransformer

Reply:
Should `Qwen2VisionTransformer` really be a `PreTrainedModel`? It's just part of the `Qwen2VLModel`,

Reply:
Sorry, I missed this comment. Yes, we will need that to be able to automatically dispatch attn implementation on vision module. And we won't add it to the high level import/docs or auto-mapping, so should be fine

Reply:
Would this pass the check_repo_consistency?

Reply:
right, it won't as it's not high-level importable but we can add the module in PRIVATE_MODELS in `utils/check_repo.py`. 

The reason we need it is because vision tower is actually a separate model, even as a part of QwenVL, so it needs to have all properties of a PreTrainedModel in loading